In [68]:
import pandas as pd
import numpy as np

In [69]:
# importing transaction dataset
df1 = pd.read_csv('QVI_transaction_data.csv')

In [70]:
# importing purchase behaviour dataset
df2 = pd.read_csv('QVI_purchase_behaviour.csv')

#### Cleaning Transaction dataset

In [71]:
df1.head()

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0  43390          1            1000       1         5   
1  43599          1            1307     348        66   
2  43605          1            1343     383        61   
3  43329          2            2373     974        69   
4  43330          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES  
0    Natural Chip        Compny SeaSalt175g         2        6.0  
1                  CCs Nacho Cheese    175g         3        6.3  
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8

#### Checking for duplicate

In [72]:
df1[df1.duplicated()]

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
124845  43374        107          107024  108462        45   

                                     PROD_NAME  PROD_QTY  TOT_SALES  
124845  Smiths Thinly Cut   Roast Chicken 175g         2        6.0

In [73]:
# droppling the duplicated row
df1 = df1.drop_duplicates()

In [74]:
df1.duplicated().sum()

0

#### Checking for the the infomation of the dataset

In [75]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 264835 entries, 0 to 264835
Data columns (total 8 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   DATE            264835 non-null  int64  
 1   STORE_NBR       264835 non-null  int64  
 2   LYLTY_CARD_NBR  264835 non-null  int64  
 3   TXN_ID          264835 non-null  int64  
 4   PROD_NBR        264835 non-null  int64  
 5   PROD_NAME       264835 non-null  object 
 6   PROD_QTY        264835 non-null  int64  
 7   TOT_SALES       264835 non-null  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 18.2+ MB


#### Changing datatype of date column

In [76]:
df1['DATE'] = pd.to_datetime(df1['DATE'], unit='D', origin='1899-12-30')

In [77]:
df1['DATE'].head()

0   2018-10-17
1   2019-05-14
2   2019-05-20
3   2018-08-17
4   2018-08-18
Name: DATE, dtype: datetime64[ns]

#### Checking for null values

In [78]:
df1.isnull().sum()

DATE              0
STORE_NBR         0
LYLTY_CARD_NBR    0
TXN_ID            0
PROD_NBR          0
PROD_NAME         0
PROD_QTY          0
TOT_SALES         0
dtype: int64

No null value, we can move on

In [79]:
df1.head()

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0 2018-10-17          1            1000       1         5   
1 2019-05-14          1            1307     348        66   
2 2019-05-20          1            1343     383        61   
3 2018-08-17          2            2373     974        69   
4 2018-08-18          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES  
0    Natural Chip        Compny SeaSalt175g         2        6.0  
1                  CCs Nacho Cheese    175g         3        6.3  
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8

#### Removing product size from product name

In [80]:
df1['PROD_NAME'].unique()

array(['Natural Chip        Compny SeaSalt175g',
       'CCs Nacho Cheese    175g',
       'Smiths Crinkle Cut  Chips Chicken 170g',
       'Smiths Chip Thinly  S/Cream&Onion 175g',
       'Kettle Tortilla ChpsHny&Jlpno Chili 150g',
       'Old El Paso Salsa   Dip Tomato Mild 300g',
       'Smiths Crinkle Chips Salt & Vinegar 330g',
       'Grain Waves         Sweet Chilli 210g',
       'Doritos Corn Chip Mexican Jalapeno 150g',
       'Grain Waves Sour    Cream&Chives 210G',
       'Kettle Sensations   Siracha Lime 150g',
       'Twisties Cheese     270g', 'WW Crinkle Cut      Chicken 175g',
       'Thins Chips Light&  Tangy 175g', 'CCs Original 175g',
       'Burger Rings 220g', 'NCC Sour Cream &    Garden Chives 175g',
       'Doritos Corn Chip Southern Chicken 150g',
       'Cheezels Cheese Box 125g', 'Smiths Crinkle      Original 330g',
       'Infzns Crn Crnchers Tangy Gcamole 110g',
       'Kettle Sea Salt     And Vinegar 175g',
       'Smiths Chip Thinly  Cut Original 175g', 'K

#### Creating a new column that will show the size of the product in g

In [81]:
df1['WEIGHTS(g)'] = df1['PROD_NAME'].apply(lambda x : x[-4:])

In [82]:
df1.head()

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0 2018-10-17          1            1000       1         5   
1 2019-05-14          1            1307     348        66   
2 2019-05-20          1            1343     383        61   
3 2018-08-17          2            2373     974        69   
4 2018-08-18          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES WEIGHTS(g)  
0    Natural Chip        Compny SeaSalt175g         2        6.0       175g  
1                  CCs Nacho Cheese    175g         3        6.3       175g  
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9       170g  
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0       175g  
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8       150g

In [83]:
df1['WEIGHTS(g)'].unique()

array(['175g', '170g', '150g', '300g', '330g', '210g', '210G', '270g',
       '220g', '125g', '110g', '134g', '150G', '175G', '380g', '180g',
       '165g', 'Salt', '250g', '200g', '160g', '190g', ' 90g', ' 70g'],
      dtype=object)

Since some cells in the column end with 'salt' and not the exact weight of the product, i have to check the cells to see what it contains

In [84]:
df1['PROD_NAME'][df1['PROD_NAME'].str.endswith('Salt')]

65        Kettle 135g Swt Pot Sea Salt
153       Kettle 135g Swt Pot Sea Salt
174       Kettle 135g Swt Pot Sea Salt
177       Kettle 135g Swt Pot Sea Salt
348       Kettle 135g Swt Pot Sea Salt
                      ...             
264564    Kettle 135g Swt Pot Sea Salt
264574    Kettle 135g Swt Pot Sea Salt
264725    Kettle 135g Swt Pot Sea Salt
264767    Kettle 135g Swt Pot Sea Salt
264823    Kettle 135g Swt Pot Sea Salt
Name: PROD_NAME, Length: 3257, dtype: object

3257 columns end with salt, however, the product size can be found inbetween the cells

#### Futher extracting weight from product name

In [85]:
def extract_weight(x):
    if x == 'Salt' :
        return '135g'
    else :
        return x
    

In [86]:
df1['WEIGHTS(g)'] = df1['WEIGHTS(g)'].apply(extract_weight)

In [87]:
df1['WEIGHTS(g)'] = df1['WEIGHTS(g)'].str[:-1] #Removing the 'G' and 'g' signifying gram at the end of the cells

In [88]:
df1['WEIGHTS(g)'].unique()

array(['175', '170', '150', '300', '330', '210', '270', '220', '125',
       '110', '134', '380', '180', '165', '135', '250', '200', '160',
       '190', ' 90', ' 70'], dtype=object)

#### Cleaning prod_name column by removing unneccessary strings and weight values from the column

In [89]:
df1['PROD_NAME']

0           Natural Chip        Compny SeaSalt175g
1                         CCs Nacho Cheese    175g
2           Smiths Crinkle Cut  Chips Chicken 170g
3           Smiths Chip Thinly  S/Cream&Onion 175g
4         Kettle Tortilla ChpsHny&Jlpno Chili 150g
                            ...                   
264831     Kettle Sweet Chilli And Sour Cream 175g
264832               Tostitos Splash Of  Lime 175g
264833                    Doritos Mexicana    170g
264834     Doritos Corn Chip Mexican Jalapeno 150g
264835               Tostitos Splash Of  Lime 175g
Name: PROD_NAME, Length: 264835, dtype: object

In [90]:
df1['PROD_NAME']= df1['PROD_NAME'].str.replace('/', '')

In [91]:
df1['PROD_NAME']= df1['PROD_NAME'].str.replace('135g ', '')

In [92]:
df1['PROD_NAME']= df1['PROD_NAME'].str.replace('S/', ' ')

In [93]:
df1['PROD_NAME'] = df1['PROD_NAME'].str.replace('\d+', ' ')

C:\Users\Temitope\AppData\Local\Temp\ipykernel_7768\3580492188.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df1['PROD_NAME'] = df1['PROD_NAME'].str.replace('\d+', ' ')


In [94]:
df1['PROD_NAME'] = df1['PROD_NAME'].str.replace('[gG]$', '')

C:\Users\Temitope\AppData\Local\Temp\ipykernel_7768\3811155919.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df1['PROD_NAME'] = df1['PROD_NAME'].str.replace('[gG]$', '')


#### Saving the cleaned transaction dataset

In [95]:
df1.to_csv('cleaned_transaction_dataset.csv', index=False)

#### Cleaning Purchase behaviour dataset

In [96]:
df2.head()

LYLTY_CARD_NBR               LIFESTAGE PREMIUM_CUSTOMER
0            1000   YOUNG SINGLES/COUPLES          Premium
1            1002   YOUNG SINGLES/COUPLES       Mainstream
2            1003          YOUNG FAMILIES           Budget
3            1004   OLDER SINGLES/COUPLES       Mainstream
4            1005  MIDAGE SINGLES/COUPLES       Mainstream

In [97]:
df2['LIFESTAGE']=df2['LIFESTAGE'].astype('category')
df2['PREMIUM_CUSTOMER']=df2['PREMIUM_CUSTOMER'].astype('category')

#### Checking for null values

In [98]:
df2.isnull().sum()

LYLTY_CARD_NBR      0
LIFESTAGE           0
PREMIUM_CUSTOMER    0
dtype: int64

No null values, we can move on

#### Checking for duplicates

In [99]:
df2[df2.duplicated()]

Empty DataFrame
Columns: [LYLTY_CARD_NBR, LIFESTAGE, PREMIUM_CUSTOMER]
Index: []

No duplicated rows, we can move on

#### Checking the infomation of the dataset

In [100]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72637 entries, 0 to 72636
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   LYLTY_CARD_NBR    72637 non-null  int64   
 1   LIFESTAGE         72637 non-null  category
 2   PREMIUM_CUSTOMER  72637 non-null  category
dtypes: category(2), int64(1)
memory usage: 709.9 KB


## Now that boths datasets are clean, we can now merge them together

Since boths dataset has the loyalty card (LYLTY_CARD_NBR) column in common, we can merge boths dataset on it.

In [101]:
df = pd.merge(df1, df2, on='LYLTY_CARD_NBR')

In [102]:
df.head()

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0 2018-10-17          1            1000       1         5   
1 2019-05-14          1            1307     348        66   
2 2018-11-10          1            1307     346        96   
3 2019-03-09          1            1307     347        54   
4 2019-05-20          1            1343     383        61   

                             PROD_NAME  PROD_QTY  TOT_SALES WEIGHTS(g)  \
0  Natural Chip        Compny SeaSalt          2        6.0        175   
1                CCs Nacho Cheese              3        6.3        175   
2          WW Original Stacked Chips           2        3.8        160   
3                       CCs Original           1        2.1        175   
4  Smiths Crinkle Cut  Chips Chicken           2        2.9        170   

                LIFESTAGE PREMIUM_CUSTOMER  
0   YOUNG SINGLES/COUPLES          Premium  
1  MIDAGE SINGLES/COUPLES           Budget  
2  MIDAGE SINGLES/COUPLES           Budget  
3  MIDAGE SINGLES/COUPLES           Budget  
4  MIDAGE SINGLES/COUPLES           Budget

In [103]:
df.shape

(264835, 11)

#### Saving the cleaned merged dataset

In [104]:
df.to_csv('Cleaned_Quantium_Merged_Dataset.csv', index=False)

#### Since i am meant to analyse the customer behaviour on the purchase of chips alone, i am going to filter the product name with goods that has chips in them. However, i am going to rename some cells because i noticed that some words are abbreviated, while some are not spelt in full.

In [105]:
def rename(PROD_NAME):
    abbreviations = {'Chps' : 'Chips', 'Chp' : 'Chips', 'Compny' : 'Company'}
    for i in abbreviations.keys():
        if i in PROD_NAME:
            PROD_NAME = PROD_NAME.replace(i, abbreviations[i])
    return PROD_NAME       

In [106]:
df['PROD_NAME'] = df['PROD_NAME'].apply(rename)

#### Now, i would filter the dataset to show only product names that has chips in it 

In [107]:
pick_chips = ' Chips'
Chips_filter = df['PROD_NAME'].str.contains(pick_chips)

In [108]:
x = Chips_filter.value_counts()

In [109]:
x

False    202300
True      62535
Name: PROD_NAME, dtype: int64

This means that there are 62,535 chips product, gotten from the product name

In [110]:
chips_product = df.loc[Chips_filter, ['DATE', 'STORE_NBR', 'LYLTY_CARD_NBR', 'TXN_ID', 'PROD_NBR', 'PROD_NAME', 'PROD_QTY', 'TOT_SALES', 'WEIGHTS(g)', 'LIFESTAGE', 'PREMIUM_CUSTOMER']]

In [111]:
chips_product

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
2      2018-11-10          1            1307     346        96   
4      2019-05-20          1            1343     383        61   
6      2018-08-18          2            2426    1038       108   
9      2018-08-09          4            4074    2979        60   
10     2018-12-12          4            4074    2980         4   
...           ...        ...             ...     ...       ...   
264817 2018-11-04        271          271193  269365        33   
264818 2019-05-29        271          271193  269366        75   
264824 2019-03-13        272          272193  269906         9   
264826 2019-03-25        272          272194  269908        75   
264829 2018-11-12        272          272319  270087        44   

                                     PROD_NAME  PROD_QTY  TOT_SALES  \
2                  WW Original Stacked Chips           2        3.8   
4          Smiths Crinkle Cut  Chips Chicken           2        2.9   
6       Kettle Tortilla ChipsHny&Jlpno Chili           3       13.8   
9           Kettle Tortilla ChipsFeta&Garlic           2        9.2   
10             Dorito Corn Chips     Supreme           2       13.0   
...                                        ...       ...        ...   
264817     Cobs Popd SwtChlli &SrCream Chips           2        7.6   
264818             Cobs Popd Sea Salt  Chips           2        7.6   
264824   Kettle Tortilla ChipsBtroot&Ricotta           1        4.6   
264826             Cobs Popd Sea Salt  Chips           2        7.6   
264829             Thins Chips Light&  Tangy           2        6.6   

       WEIGHTS(g)               LIFESTAGE PREMIUM_CUSTOMER  
2             160  MIDAGE SINGLES/COUPLES           Budget  
4             170  MIDAGE SINGLES/COUPLES           Budget  
6             150  MIDAGE SINGLES/COUPLES           Budget  
9             150  MIDAGE SINGLES/COUPLES           Budget  
10            380  MIDAGE SINGLES/COUPLES           Budget  
...           ...                     ...              ...  
264817        110   YOUNG SINGLES/COUPLES          Premium  
264818        110   YOUNG SINGLES/COUPLES          Premium  
264824        150   YOUNG SINGLES/COUPLES          Premium  
264826        110   YOUNG SINGLES/COUPLES          Premium  
264829        175   YOUNG SINGLES/COUPLES          Premium  

[62535 rows x 11 columns]

In [112]:
chips_product = chips_product.reset_index()

# EDA

Using a def function to create a new column that would show if the product bought is either small, medium or large based on the product quality 

In [113]:
def size(number):
    if number < 100:
        return 'Small'
    if number >= 100 and number < 200:
        return 'Medium'
    if number >= 200:
        return 'Large'

In [114]:
chips_product['PROD_SIZE']=chips_product['PROD_QTY'].apply(size)

Creating a column that would provide the price of the product

In [115]:
chips_product['product_price'] = chips_product['TOT_SALES'] / chips_product['PROD_QTY']

Creating a month column to help provide insight showing the hightest and lowest month sales

In [116]:
chips_product['DATE'] = chips_product['DATE'].dt.month_name()

Creating a column from the product name that would show the brand name

In [117]:
def Brand_name(a):
    b=a.split(" ")
    Brand=b[0]
    
    return Brand

In [118]:
chips_product['BRAND NAME']=chips_product['PROD_NAME'].apply(Brand_name)

Creating a column that would show the age group of the customers

In [119]:
def age_cat(a):
    b=a.split(" ")
    Age=''
    if b[0]=='NEW':
        Age='MIDAGE';
    elif b[0]=='RETIREES':
        Age='OLDER'
    else:
        Age=b[0]
    return Age

In [120]:
chips_product['AGE CATEGORY']=chips_product['LIFESTAGE'].apply(age_cat)

In [121]:
chips_product.head()

index      DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0      2  November          1            1307     346        96   
1      4       May          1            1343     383        61   
2      6    August          2            2426    1038       108   
3      9    August          4            4074    2979        60   
4     10  December          4            4074    2980         4   

                                PROD_NAME  PROD_QTY  TOT_SALES WEIGHTS(g)  \
0             WW Original Stacked Chips           2        3.8        160   
1     Smiths Crinkle Cut  Chips Chicken           2        2.9        170   
2  Kettle Tortilla ChipsHny&Jlpno Chili           3       13.8        150   
3      Kettle Tortilla ChipsFeta&Garlic           2        9.2        150   
4         Dorito Corn Chips     Supreme           2       13.0        380   

                LIFESTAGE PREMIUM_CUSTOMER PROD_SIZE  product_price  \
0  MIDAGE SINGLES/COUPLES           Budget     Small           1.90   
1  MIDAGE SINGLES/COUPLES           Budget     Small           1.45   
2  MIDAGE SINGLES/COUPLES           Budget     Small           4.60   
3  MIDAGE SINGLES/COUPLES           Budget     Small           4.60   
4  MIDAGE SINGLES/COUPLES           Budget     Small           6.50   

  BRAND NAME AGE CATEGORY  
0         WW       MIDAGE  
1     Smiths       MIDAGE  
2     Kettle       MIDAGE  
3     Kettle       MIDAGE  
4     Dorito       MIDAGE

Saving the dataset showing only chips product

In [123]:
chips_product.to_csv('Cleaned_chips_product_dataset.csv', index=False)